### DISCUSSION TITLE
mhe.set_initial_guess() doesn't seem to work

### DISCUSSION THREAD

**Discussion by shanek16** [do-mpc/do-mpc#286] [General]
Dear developers of dompc

**Describe the bug**
Following the document: https://www.do-mpc.com/en/latest/mhe_example.html
I want to start estimation from the initial observed value.
However, using the code from above link doesn't seem to work.

you can see from the image in the document:
![image](https://user-images.githubusercontent.com/53452585/182829562-f252fb38-d3e6-4732-80fd-77c26d8d1218.png)
 that estimation starts from 0, although we set the initial guess at **x0_mhe** by following code:
x0 = np.pi*np.array([1, 1, -1.5, 1, -5, 5, 0, 0]).reshape(-1,1)
x0_mhe = x0*(1+0.5*np.random.randn(8,1))
mhe.x0_mhe = x0_mhe
mhe.set_initial_guess()

**Is the code above not for setting the initial estimate where I want?
If not, what is the code designed for?**

It is strange that in the image after **MHE Advantages**:
![image](https://user-images.githubusercontent.com/53452585/182830876-774a9224-3648-484c-af21-bc2130d3d41f.png)
estimation starts from initial observed value, although we didn't set the initial estimation.

**Which version of do-mpc are using and how did you obtain it?**
installed via git clone
version:
![image](https://user-images.githubusercontent.com/53452585/182830187-27b21aa3-02f5-4c57-be7f-6263fe286934.png)

**Are you running our examples or your own code?**
- this is the code from document: https://www.do-mpc.com/en/latest/mhe_example.html

**Which system are you working on?**
- Linux

Thank you for your development always :)

**Comment by 4flixt** [comment]
Hi @shanek16 ,

so the MHE is a bit complicated in this regard. The ``mhe.x0`` is the currently estimated state and you can set it before the first iteration to create an initial guess with ``mhe.set_initial_guess()``. This does only create a guess for the optimizer, however.
Setting ``mhe.x0``doesn't change anything else.

More important for the MHE is the sequence of measurements (with finite horizon) that are supplied at each call. 
If you want to properly implement an MHE, you will have do define, how to obtain that most recent sequence of measurements with [mhe.set_yfun](https://www.do-mpc.com/en/latest/api/do_mpc.estimator.MHE.set_y_fun.html).

We have implemented a default measurement function, however, which is used in this example. This measurement function, reads from the ``mhe.data`` object the most recent ``N_horizon`` values of measurements.
This creates a problem though: If we haven't run the MHE for ``N_horzizon`` steps, not enough measurements are available. 
To fix this problem, our default measurement function extrapolates the (hold the last value). 

**Now how does this explain the issue you have pointed out?**
Well when we first run the MHE, there are no real measurements available and they are all set to zero. The MHE consequently starts from 0 (regardless of your ``x0``).

In the second example (Section "MHE Advantages"), we had already run the MHE (in the first example). The measurement equation thus returned useful measurements and we started from a non-zero initial state. 

I hope this helps!

**Reply by shanek16** [reply]
Dear @4flixt 
Pardon my late reply.
Thank you greatly for your answer and explanation about mhe.set_yfun.

I wanted to start estimation from the initial observed value from the first run of MHE, so
I tried setting the initial estimation state by using:
```
x0 = np.pi*np.array([1, 1, -1.5, 1, -5, 5, 0, 0]).reshape(-1,1)
x0_mhe = x0*(1+0.5*np.random.randn(8,1))
simulator.x0 = x0
# mhe.x0_mhe = x0_mhe
mhe.x0 = x0_mhe
mhe.p_est0 = 1e-4
mhe.set_initial_guess()

```
and it seems to work as I intended: To start estimation from the initial observed value.
![image](https://user-images.githubusercontent.com/53452585/188678757-7c678d78-6016-4123-9f2f-62f4c5d801c0.png)

Would this be an appropriate way to start estimation from the initial observed value? Or am I leaving something out?
I'm not sure if I'm doing the right thing here..
Thanks.

Sincerely, Shane K.
